python/ray/_raylet.pyx

```diff
@@ -404,6 +417,259 @@ cdef class RayletClient:
+                results.append(data)
+            else:
+                results.append(pickle.loads(data.to_pybytes()))
+        # Passed by reference.
+        else:
+            by_reference_ids.append(
+                ObjectID(arg_reference_ids[i].Binary()))
+            by_reference_indices.append(i)
+            by_reference_objects.push_back(args[i])
+            results.append(None)
+
+    data_metadata_pairs = RayObjectsToDataMetadataPairs(
+        by_reference_objects)
+    for i, result in enumerate(
+        ray.worker.global_worker.deserialize_objects(
+            data_metadata_pairs, by_reference_ids)):
+        results[by_reference_indices[i]] = result
+
+    return results
+
+cdef _check_worker_state(worker, CTaskType task_type, JobID job_id):
+    assert worker.current_task_id.is_nil()
+    assert worker.task_context.task_index == 0
+    assert worker.task_context.put_index == 1
+
+    # If this worker is not an actor, check that `current_job_id`
+    # was reset when the worker finished the previous task.
+    if <int>task_type in [<int>TASK_TYPE_NORMAL_TASK,
+                          <int>TASK_TYPE_ACTOR_CREATION_TASK]:
+        assert worker.current_job_id.is_nil()
+        # Set the driver ID of the current running task. This is
+        # needed so that if the task throws an exception, we propagate
+        # the error message to the correct driver.
+        worker.current_job_id = job_id
+    else:
+        # If this worker is an actor, current_job_id wasn't reset.
+        # Check that current task's driver ID equals the previous
+        # one.
+        assert worker.current_job_id == job_id
+
+
+cdef _store_task_outputs(worker, return_ids, outputs):
+    for i in range(len(return_ids)):
+        return_id, output = return_ids[i], outputs[i]
+        if isinstance(output, ray.actor.ActorHandle):
+            raise Exception("Returning an actor handle from a remote "
+                            "function is not allowed).")
+        if output is ray.experimental.no_return.NoReturn:
+            if not worker.core_worker.object_exists(return_id):
+                raise RuntimeError(
+                    "Attempting to return 'ray.experimental.NoReturn' "
+                    "from a remote function, but the corresponding "
+                    "ObjectID does not exist in the local object store.")
+        else:
+            worker.put_object(return_id, output)
+
+
+cdef execute_task(
+        CTaskType task_type,
+        const CRayFunction &ray_function,
+        const CActorID &c_actor_id,
+        const unordered_map[c_string, double] &c_resources,
+        const c_vector[shared_ptr[CRayObject]] &c_args,
+        const c_vector[CObjectID] &c_arg_reference_ids,
+        const c_vector[CObjectID] &c_return_ids,
+        c_vector[shared_ptr[CRayObject]] *returns):
+
+    worker = ray.worker.global_worker
+
+    actor_id = ActorID(c_actor_id.Binary())
+    job_id = worker.core_worker.get_current_job_id()
```

Comment:
I'm a little confused by the logic between here and `_check_worker_state`...isn't `_check_worker_state` supposed to take in the job ID of the current task, not `worker.core_worker.get_current_job_id()`?

Reply:
Ah, good catch!